packages/apollo-language-server/src/utilities/graphql.ts

```diff
@@ -87,31 +93,180 @@ export function removeDirectives(ast: ASTNode, directiveNames: string[]) {
   });
 }
 
-// remove fields where a given directive is found
-export function removeDirectiveAnnotatedFields(
-  ast: ASTNode,
+/**
+ * Recursively remove orphaned fragment definitions that have their names included in
+ * `fragmentNamesEligibleForRemoval`
+ *
+ * We expclitily require the fragments to be listed in `fragmentNamesEligibleForRemoval` so we only strip
+ * fragments that were orphaned by an operation, not fragments that started as oprhans
+ */
+function removeOrphanedFragmentDefinitions<AST extends ASTNode>(
+  ast: AST,
+  fragmentNamesEligibleForRemoval: Set<string>
+): AST {
+  /**
+   * Flag to keep track of removing any fragments
+   */
+  let anyFragmentsRemoved = false;
+
+  // Aquire names of all fragment spreads
+  const fragmentSpreadNodeNames = new Set<string>();
+  visit(ast, {
+    FragmentSpread(node) {
+      fragmentSpreadNodeNames.add(node.name.value);
+    }
+  });
+
+  // Strip unused fragment definitions. Flag if we've removed any so we know if we need to continue
+  // recursively checking.
+  ast = visit(ast, {
+    FragmentDefinition: {
+      enter(node) {
+        if (
+          fragmentNamesEligibleForRemoval.has(node.name.value) &&
+          !fragmentSpreadNodeNames.has(node.name.value)
+        ) {
+          // This definition is not used, remove it.
+          anyFragmentsRemoved = true;
+          return null;
+        }
+
+        return undefined;
+      }
+    },
+    FragmentSpread: {
+      leave(node) {
+        if (!fragmentSpreadNodeNames.has(node.name.value)) {
+          // This definition is not used, remove it.
+          anyFragmentsRemoved = true;
+          return null;
+        }
+
+        return undefined;
+      }
+    }
+  });
+
+  if (anyFragmentsRemoved) {
+    // We've removed fragments and might have orphaned more fragments, so recursively try to remove more
+    // orphaned fragments.
+    return removeOrphanedFragmentDefinitions(
+      ast,
+      fragmentNamesEligibleForRemoval
+    );
+  }
+
+  return ast;
+}
+
+/**
+ * Recursively remove nodes that have zero-length selection sets
+ */
+function removeNodesWithEmptySelectionSets<AST extends ASTNode>(ast: AST): AST {
+  /**
+   * Flag to know if we need to make another recursive pass
+   */
+  let anyNodesRemoved = false;
+
+  ast = visit(ast, {
+    enter(node) {
+      // If this node _has_ a `selectionSet` and it's zero-length, then remove it.
+      if (
+        "selectionSet" in node &&
+        node.selectionSet != null &&
+        node.selectionSet.selections.length === 0
+      ) {
+        anyNodesRemoved = true;
+        return null;
+      }
+
+      return undefined;
+    }
+  });
+
+  if (anyNodesRemoved) {
```

Comment:
This recursion shouldn't be needed since the visitor will recurse for you going top down. As soon as a node should be trimmed, it *can't* have any children to need to be trimmed